doc/how_to/migrate/from_streamlit.md

```diff
@@ -0,0 +1,1011 @@
+```
+
+![Panel Cache Example](https://assets.holoviz.org/panel/gifs/panel_cache_example.gif)
+
+You can also use `pn.cache` as an function. I.e. as
+
+```python
+plot = pn.cache(plot)
+```
+
+Using `pn.cache` as a function can help you keep your business logic
+(`data` and `plot` function) and your caching logic (when and how to apply caching) separate. This
+can help you reusable and maintainable code.
+
+### Typical Caching Mistakes
+
+- Using `pn.state.cache` instead of `pn.cache`. `pn.state.cache` is a simple dictionary and cannot be used as a function or a decorator.
+
+### Cache Migration Steps
+
+To migrate
+
+- replace `st.cache_data` and `st.cache_resource` with `pn.cache`.
+
+You should consider caching expensive
+
+- functions that are run when your page loads
+- *bound functions*
+
+## Multi Page Apps
+
+Migrating your Streamlit multi page app to Panel is simple. In Panel each page is simply a file
+that you *serve*
+
+```bash
+panel serve home.py page1.py page2.ipynb
+```
+
+You can specify the *home* page with the `--index` flag.
+
+```bash
+panel serve home.py page1.py page2.ipynb --index=home
+```
+
+## Session State
+
+Session state (and callbacks) were added to Streamlit as an after thought
+to enable sharing variables between re-runs and across apps inside a multipage app.
+
+Panels execution model is very different and session state is not really a topic.
+
+### Streamlit Session State Example
+
+```python
+import random
+from uuid import uuid4
+
+import streamlit as st
+
+st.session_state.setdefault("rows", {})
+
+def create_row(hex_id=None):
+    if hex_id is None:
+        hex_id = uuid4().hex
+        value = random.randint(0, 100)
+        st.session_state.rows[hex_id] = value
+    else:
+        value = st.session_state.rows[hex_id]
+
+    st.number_input("Enter a number", value=value, key=f"num_{hex_id}")
+
+st.title("Dynamically add new rows")
+
+button_cols = st.columns(2)
+with button_cols[0]:
+    add_row = st.button("Add new row", use_container_width=True)
+with button_cols[1]:
+    clear_rows = st.button("Clear all rows", use_container_width=True)
+
+if clear_rows:
+    st.session_state.rows = {}
+
+for hex_id in st.session_state.rows.keys():
+    create_row(hex_id=hex_id)
+
+if add_row:
+    create_row()
+```
+
+![Streamlit Session State Example](https://user-images.githubusercontent.com/42288570/248444403-3e9548a5-71a6-4154-8e85-b9d9e08b9111.png)
+
+With Streamlit you need session state to keep track of the dynamically created `rows`.
+
+### Panel Session State Example
+
+```python
+import random
+
+import panel as pn
+pn.extension(sizing_mode="stretch_width")
+
+def create_row(_):
+    value = random.randint(0, 100)
+    row = pn.widgets.TextInput(name="Enter a number", value=str(value))
+    rows.append(row)
+
+rows = pn.Column()
+add_row = pn.widgets.Button(name="Add new row", on_click=create_row)
+clear_rows = pn.widgets.Button(name="Clear all rows", on_click=lambda _: rows.clear())
```

Comment:
Thanks. I am a little hesitant to comitting because this version has the issue of using the `_` as a function argument. I don't think beginners are familiar with that either. I've never been working with any python framework that is creating functions with an `_`. Its just so hard to explain.

Reply:
What do you think @philippjfr . Is it better? For now?

Reply:
No I think it should be called `event`, even if it's unused.

Reply:
Okay updated the code suggestion.

Reply:
Explanatory material is not production code and does not have to follow strict linting rules. It is also not weird or unreasonable for an `on_click` handler to receive events, e.g. see:

- Dash: https://dash.plotly.com/dash-html-components/button
- ipywidgets: https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html#example

Not objecting to having both be supported by inspecting the call signature but there's also such a thing as underestimating users and I'd consider the friction here very minor.

Reply:
The problem is also that very often the function is coming from some Python framework or some code you have created before anyone though about Panel and UIs. Why can I not just run that function when clicking a button. Why do I have to wrap it in a function or lambda? Its friction.